transaction/core/src/blockchain/block_version.rs

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2018-2021 The MobileCoin Foundation
+
+use core::{fmt, hash::Hash, ops::Deref};
+use mc_crypto_digestible::Digestible;
+use serde::{Deserialize, Serialize};
+
+/// A block version number
+#[derive(
+    Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize, Digestible, Hash,
+)]
+pub struct BlockVersion(pub u32);
+
+impl From<u32> for BlockVersion {
+    fn from(src: u32) -> Self {
+        Self(src)
+    }
+}
```

Comment:
This should be `TryFrom<u32>`, not `From<u32>`

This would gate supported versions from unsupported, rather than having to re-implement that test everywhere you're reading an integer. If you want an infallible constructor, it should be `impl BlockVersion { fn from_unchecked(src: u32) -> Self { Self(src) } }`

```suggestion
pub struct UnsupportedVersion;

impl TryFrom<u32> for BlockVersion {
    type Err = UnsupportedVersion;

    fn try_from(src: u32) -> Result<Self, Self::Err> {
        if src < Self::MIN.0 || src > Self::MAX.0 {
            Err(UnsupportedVersion)
        } else {
            Ok(Self(src))
        }
    }
}
```

Reply:
This has some drawbacks: for instance, in Eran's idea: https://github.com/mobilecoinfoundation/mobilecoin/pull/1461#pullrequestreview-876465436

We might tell people about block versions that are coming soon before they actually happen. If that manifests as a deserialization error that would make it harder for the app to do the right thing.

Reply:
Also this wouldn't actually work for e.g. the enclave checking this number when it is passed in, because it uses serde deserialize and not try from. So it's potentially a footgun to do this check "implicitly" as you suggest. I feel like it's safer to check the number against max everywhere that it's important to do so

Reply:
I'm okay to do it this way if you want, I'll have to figure out a nicer way to iterate over block versions though

Reply:
Actually it was the iteration that led me to this, hence `BlockVersion::MIN..BlockVersion::MAX`

Reply:
Serde is the good point, but I'd think that `BlockVersion::is_valid()` is the better work-around for the lack of proper data-access serializable objects.

Reply:
For the comment in #1461, I think this is fine, that's a good "time to upgrade" indicator (when `try_from()` fails).

Reply:
This might get ugly in various protobuf<->prost conversions, if we have them. E.g. I am imagining a world where some grpc (protobuf) response returns a bunch of fields and we try to convert it into our Prosty struct, but now this fails because the struct has a block_version in it and the `try_from` fails. We could use the `from_unchecked` variant there, but then that kind of means we can't rely on whats inside `BlockVersion` to be valid.

Separately, I am not sure the concept of valid/invalid makes sense here. One way I am thinking about it is all block versions are valid, we might just not know what to do with a given version yet so it is unsupported for certain purposes (like, we cant hash it, and we cant build it)

Reply:
@jcape another alternative may be, `BlockVersion` objects are not valid and invalid, but attempting to determine if a feature is supported looks like:

```
pub fn feature_memos_is_supported(&self) -> Result<bool> {
    if self > BlockVersion::MAX { return Err(...); }
    Ok(self.0 >= 2)
}
```

so using it for transaction building or validation fails if this version is too new and you don't know what it the rules are, but deserializing a `BlockVersion` doesn't fail.